### PR TITLE
Move to dotnet cli for building, and .NET Core (netstandard1.0) as target platform

### PR DESCRIPTION
**Pull request by jskeet**
This is only the changes under csharp. Next I need to change the Jenkins, Travis and AppVeyor builds.

I've fetched this from clean on an Ubuntu 16.04 box, and it builds and tests pass.
@jtattermusch: Assigning to you for initial comments, while I start looking at the CI aspects. We're not committed to this just yet - if we decide it's too risky at this stage, we can abandon it.

### CODE REVIEW

File: csharp/src/Google.Protobuf/project.json [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    "Release": {
+      "buildOptions": {
+        "define": [ "RELEASE", "TRACE" ],
+        "optimize": true
+      }
+    }
+  },
+
+  "frameworks": {
+    // TODO: A net45 target here? Need to check that targeting this
```
Comment (jtattermusch) [inline csharp/src/Google.Protobuf/project.json]: If adding net45 or profile259 here prevents the need to upgrade to nuget 2.12 with VS2013, I am all in favor of this. (protobuf is supposed to be highly portable so we should try to impose as few restrictions on user's environment as possible).

Comment (jskeet) [inline csharp/src/Google.Protobuf/project.json]: Hmm. Would personally rather not do that, but let's get everything else done first. Will leave the TODO for the moment, at least.

**Comment by jskeet** [comment]
- AppVeyor tests now pass
- Travis will take a bit longer to work out (due to the very slow iteration cycle)
- Jenkins confuses me; the tests run, but there's an error in the overall build:

Failed to read test report file /home/jenkins/workspace/protobuf_pull_request/testoutput/testresults.xml
org.dom4j.DocumentException: Error on line 2225 of document file:///home/jenkins/workspace/protobuf_pull_request/testoutput/testresults.xml : An invalid XML character (Unicode: 0x8) was found in the element content of the document. Nested exception: An invalid XML character (Unicode: 0x8) was found in the element content of the document.

I don't understand what's going on there... @jtattermusch do you have enough Jenkins experience to know what's going on?

**Comment by pgrosu** [comment]
Hi Jon,

So based on the [console output](https://grpc-testing.appspot.com/job/protobuf_pull_request/296/consoleText) at the end, the following post-processing is happening:

```
+ COPY_FROM_DOCKER=/var/local/git/protobuf/testoutput
+ mkdir -p /var/local/git/protobuf/testoutput
+ TESTOUTPUT_XML_FILE=/var/local/git/protobuf/testoutput/testresults.xml
+ python /var/local/jenkins/protobuf/jenkins/make_test_output.py /tmp/tmp.TD3Lrxo0Vk
make_test_output.py: writing XML from directory: /tmp/tmp.TD3Lrxo0Vk
+ ls -l /var/local/git/protobuf/testoutput/testresults.xml
-rw-r--r-- 1 root staff 2555545 Jul  1 08:41 /var/local/git/protobuf/testoutput/testresults.xml
```

So maybe after running [make_test_output.py](https://github.com/google/protobuf/blob/master/jenkins/make_test_output.py) to bundle the tests together into an XML file - assuming it is structured properly after generating through [yattag](https://pypi.python.org/pypi/yattag) - it might help to have a script containing `sed` commands for post-processing out any invalid XML characters, since the tests might contain those characters.  The one below will remove the backspace (`0x8`) character:

```
#!/bin/bash
ORIGINAL_PWD=`pwd`
cd /var/local/git/protobuf/testoutput/
BACKSPACE=$(printf '\b')
sed -e "s/$BACKSPACE//g" testresults.xml > processed_testresults.xml
cp processed_testresults.xml testresults.xml
rm processed_testresults.xml
cd $ORIGINAL_PWD
```

Or this can be performed on all the tests before being bundled into an XML file.  Maybe Jan might have a better approach.

Hope it helps,
Paul

**Comment by jskeet** [comment]
Thanks Paul. Trying just adding a simple sed after the python call. I'm puzzled as to where these backspaces are coming from though...

**Comment by jskeet** [comment]
Okay, looking at https://grpc-testing.appspot.com/job/protobuf_pull_request/300/testReport/(root)/csharp/csharp/ it looks like it's a problem due to the first run of a .NET Core environment doing some pre-work that includes backspaces.

Will see if I can get rid of that...

**Comment by pgrosu** [comment]
The Jenkins plugin seems to be reporting the test results properly now:

https://grpc-testing.appspot.com/job/protobuf_pull_request/301/testReport/(root)/

Thanks for the fix,
~p

**Comment by jskeet** [comment]
@pgrosu: Yup - it was basically a case of doing a "dummy run" and deliberately swallowing the output first. Ugly but effective :)

**Comment by jskeet** [comment]
Reassigned to @pherl is @jtattermusch is away. The Travis build failures are the ones afflicting the whole project, not just this PR.

**Comment by jskeet** [comment]
@chrisdunelm - could you review this, as @jtattermusch is unavailable at the moment? I can walk you through any bits you're unfamiliar with.

**Comment by jskeet** [comment]
Note to self: I want to build the package and test it locally for dependencies when using it from .NET 4.5.

**Comment by jskeet** [comment]
Just checked, and we definitely need to either have a net45 target or change the dependencies to netstandardlibrary1.6. Will look at that in the morning.

**Comment by jskeet** [comment]
(Jenkins failure is for Python.)

@jtattermusch: Addressing a comment you made elsewhere, I think it's fine to require PCL users to update to a recent version of nuget. I completely agree that protobuf should be as portable as possible, but the easiest way of making that portability maintainable is to require that the *developers* (rather than end users) are using recent versions of tool components. I've added a `net45` target framework to the package to avoid redundant dependencies being requested, but I don't think it's worth doing the same for the portable-* version, when VS users will be fine if they update to nuget 2.12.

**Comment by chrisdunelm** [comment]
LGTM

**Comment by jskeet** [comment]
Travis failure for cocoapods is unrelated. Merging (finally!)

**Pull request merged**